### PR TITLE
Introduce an organizeImports command

### PR DESCRIPTION
**Pull request by amcasey**
In phase 1, it coalesces imports from the same module and sorts the
results, but does not remove unused imports.

Some trivia is lost during coalescing, but none should be duplicated.

### CODE REVIEW

**Comment by amcasey** [comment]
This will need to be reconciled with https://github.com/Microsoft/TypeScript/pull/21876.

File: src/services/services.ts [hunk 1]
```diff
@@ -2327,4 +2380,238 @@ namespace ts {
...
+                }
+            }
+
+            for (const namedImport of namedImports) {
+                for (const specifier of namedImport.elements) {
+                    newImportSpecifiers.push(specifier);
+                }
+            }
+
+            const sortedImportSpecifiers = stableSort(newImportSpecifiers, (s1, s2) => {
```
Comment (amcasey) [inline src/services/services.ts]: This comparison should be shared with https://github.com/Microsoft/TypeScript/pull/21876.

File: src/services/services.ts [hunk 2]
```diff
@@ -1907,6 +1907,58 @@ namespace ts {
...
+
+            // All of the (old) ImportDeclarations in the file, in syntactic order.
+            const oldImportDecls: ImportDeclaration[] = [];
+
+            forEachChild(sourceFile, node => {
+                cancellationToken.throwIfCancellationRequested();
+                if (isImportDeclaration(node)) {
+                    oldImportDecls.push(node);
+                }
+                // TODO (https://github.com/Microsoft/TypeScript/issues/10020): sort *within* ambient modules (find using isAmbientModule)
```
Comment (amcasey) [inline src/services/services.ts]: Oops, forgot to mention this other TODO in the description.

**Comment by amcasey** [comment]
I don't understand the build failures.  I ran `jake runtests-parallel` locally and it claims to have run the same link command as the server.  It reported no errors.  However, when I ran the same lint command manually, its output matched the Jenkins console output.  Is it possible that `runtests-parallel` is swallowing (or not waiting for) the lint errors? /cc @weswigham

**Comment by amcasey** [comment]
```
e:\ts_gh>jake runtests-parallel
Discovered 100 unittest suites.
Discovering runner-based tests...
Discovered 11829 test files in 504ms.
Starting to run tests using 8 threads...
Batching initial test lists...
Batched into 8 groups with approximate total time of 1m34s in each group. (90.0% of total tests batched)

  [▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬▬] √ 59191 passing (2m9s)


  59191 passing (2m)

Linting: node node_modules/tslint/bin/tslint Gulpfile.ts scripts/generateLocalizedDiagnosticMessages.ts "scripts/tslint/**/*.ts" "src/**/*.ts" --exclude "src/lib/*.d.ts" --formatters-dir ./built/local/tslint/formatters --format autolinkableStylish
rm -rf tests\baselines\local\projectOutput\

e:\ts_gh>
```

**Comment by weswigham** [comment]
I think our mistake is that windows shells don't expand globbing patterns natively? Probably? I'm not sure.

File: src/harness/unittests/organizeImports.ts [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+                    const sl2 = node2 as LiteralLikeNode;
+                    assert.equal(sl1.text, sl2.text);
+                    break;
+                default:
+                    assert.equal(node1.getText(), node2.getText());
+                    break;
+            }
+        }
+
+        function assertListEqual(list1: ReadonlyArray<Node>, list2: ReadonlyArray<Node>) {
```
Comment (andy-ms) [inline src/harness/unittests/organizeImports.ts]: I think it would be simpler to emit and compare strings, vs parsing and comparing ASTs. That way there's no need to write `assertEqual`  and we also test if we are correctly formatting the result. Also would make error messages more readable.

Comment (amcasey) [reply src/harness/unittests/organizeImports.ts]: Suits me.  How do I stringify them?

Comment (andy-ms) [reply src/harness/unittests/organizeImports.ts]: I'm attempting to do this, but the emitter works best if the ast it's working on is either *entirely* synthesized, or not at all. I'm debugging and it looks like there are some nodes with `pos` set  as children of a synthesized node, which causes the emitter to look at the source file for comments, only there is no source text. Some part of organizeImports might be forgetting to clone a node?

File: src/services/services.ts [hunk 4]
```diff
@@ -1907,6 +1907,58 @@ namespace ts {
             return codefix.getAllFixes({ fixId, sourceFile, program, host, cancellationToken, formatContext });
         }
 
+        function organizeImports(scope: OrganizeImportsScope, formatOptions: FormatCodeSettings): ReadonlyArray<FileTextChanges> {
```
Comment (andy-ms) [inline src/services/services.ts]: I think these changes are substantial enough to deserve an `organizeImports.ts` file, just like we have for breakpoints/classifier/etc.

File: src/harness/unittests/organizeImports.ts (hunk 3 above)
Comment (amcasey) [reply src/harness/unittests/organizeImports.ts]: Regarding testing with text, it seems not to have been simpler.  I'm going to leave things as they are until we have a concrete problem to correct or improvement to make.

I was under the impression that the baseline tests were doing an end-to-end test.  Did you mean something else?

File: src/services/organizeImports.ts [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        return changes;
+    }
+
+    function removeUnusedImports(oldImports: ReadonlyArray<ImportDeclaration>) {
+        return oldImports; // TODO (https://github.com/Microsoft/TypeScript/issues/10020)
+    }
+
+    /* @internal */ // Internal for testing
+    export function sortImports(oldImports: ReadonlyArray<ImportDeclaration>) {
+        if (oldImports.length < 2) {
```
Comment (andy-ms) [inline src/services/organizeImports.ts]: * In `core.ts`:
```ts
export function compareBooleans(a: boolean, b: boolean): Comparison {
    return compareValues(a ? 1 : 0, b ? 1 : 0);
}
```
* Perform `length < 2` optimization inside `stableSort`
* Here:
```ts
return stableSort(oldImports, ({ moduleSpecifier: m1 }, { moduleSpecifier: m2 }) => {
    const name1 = getExternalModuleName(m1) || m1.getText();
    const name2 = getExternalModuleName(m2) || m2.getText();
    return compareBooleans(isExternalModuleNameRelative(name1), isExternalModuleNameRelative(name2)) || compareStringsCaseSensitive(name1, name2);
});
```
This would break the `non-relative vs invalid` and `relative vs invalid` tests, but I don't think it really matters how we sort invalid expressions in import positions.

File: src/services/organizeImports.ts [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+                oldImportDecls.push(node);
+            }
+            // TODO (https://github.com/Microsoft/TypeScript/issues/10020): sort *within* ambient modules (find using isAmbientModule)
+        });
+
+        if (oldImportDecls.length === 0) {
+            return [];
+        }
+
+        const usedImportDecls = removeUnusedImports(oldImportDecls);
```
Comment (RyanCavanaugh) [inline src/services/organizeImports.ts]: Probably too many calls to `throwIfCancellationRequested` here - in general you only need to call this every ~50-200ms, and I don't think `sortImports` or `coalesceImports` could ever be that slow.

File: src/services/organizeImports.ts [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+                newDefaultImport = defaultImports[0];
+            }
+            else {
+                for (const defaultImport of defaultImports) {
+                    newImportSpecifiers.push(
+                        createImportSpecifier(createIdentifier("default"), defaultImport));
+                }
+            }
+
+            for (const namedImport of namedImports) {
```
Comment (andy-ms) [inline src/services/organizeImports.ts]: `newImportSpecifiers.push(...flatMap(namedImports, n => n.elements));`

File: src/services/organizeImports.ts [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+        if (sortedImports.length === 0) {
+            return sortedImports;
+        }
+
+        const coalescedImports: ImportDeclaration[] = [];
+
+        const groupedImports = groupSortedImports(sortedImports);
+        for (const importGroup of groupedImports) {
+
+            let seenImportWithoutClause = false;
```
Comment (andy-ms) [inline src/services/organizeImports.ts]: Think this could be its own function:
```ts
function getImportParts(importGroup: ReadonlyArray<ImportDeclaration>
): {
    importWithoutClause: ImportDeclaration | undefined,
    defaultImports: ReadonlyArray<Identifier>,
    namespaceImports: ReadonlyArray<NamespaceImport>,
    namedImports: ReadonlyArray<NamedImports>,
} {
    let importWithoutClause: ImportDeclaration;
    const defaultImports: Identifier[] = [];
    const namespaceImports: NamespaceImport[] = [];
    const namedImports: NamedImports[] = [];

    for (const importDeclaration of importGroup) {
        if (importDeclaration.importClause === undefined) {
            // Only the first such import is interesting - the others are redundant.
            // Note: Unfortunately, we will lose trivia that was on this node.
            importWithoutClause = importDeclaration;
            continue;
        }

        const { name, namedBindings } = importDeclaration.importClause;

        if (name) {
            defaultImports.push(name);
        }

        if (namedBindings) {
            if (isNamespaceImport(namedBindings)) {
                namespaceImports.push(namedBindings);
            }
            else {
                namedImports.push(namedBindings);
            }
        }
    }

    return { importWithoutClause, defaultImports, namespaceImports, namedImports };
}
```
then:
```ts
const { importWithoutClause, defaultImports, namedImports, namespaceImports } = getImportParts(importGroup);
if (importWithoutClause) {
    coalescedImports.push(importWithoutClause);
}
...proceed as before...
```

File: src/services/organizeImports.ts [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+                }
+            }
+
+            for (const namedImport of namedImports) {
+                for (const specifier of namedImport.elements) {
+                    newImportSpecifiers.push(specifier);
+                }
+            }
+
+            const sortedImportSpecifiers = stableSort(newImportSpecifiers, (s1, s2) => {
```
Comment (andy-ms) [inline src/services/organizeImports.ts]:
```ts
stableSort(newImportSpecifiers, (s1, s2) =>
    compareIdentifiers(s1.propertyName || s1.name, s2.propertyName || s2.name) || compareIdentifiers(s1.name, s2.name));
```

File: src/services/organizeImports.ts [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+                // Drop the name, if any
+                coalescedImports.push(
+                    updateImportDeclarationAndClause(namespaceImport.parent, /*name*/ undefined, namespaceImport));
+            }
+
+            if (defaultImports.length === 0 && namedImports.length === 0) {
+                continue;
+            }
+
+            let newDefaultImport: Identifier = undefined;
```
Comment (andy-ms) [inline src/services/organizeImports.ts]: `Identifier | undefined` not `Identifier = undefined`

File: src/harness/unittests/organizeImports.ts [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+            it("Combine side-effect-only imports", () => {
+                const sortedImports = parseImports(
+                    `import "lib";`,
+                    `import "lib";`);
+                const actualCoalescedImports = OrganizeImports.coalesceImports(sortedImports);
+                const expectedCoalescedImports = parseImports(`import "lib";`);
+                assertListEqual(expectedCoalescedImports, actualCoalescedImports);
+            });
+
+            it("Combine namespace imports", () => {
```
[the remark this answers is not in the document]
Comment (amcasey) [reply src/harness/unittests/organizeImports.ts]: The name describes the action taken, rather than the outcome.  Combining namespace imports happens to be a no-op.

File: src/harness/unittests/organizeImports.ts [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+                const sortedImports = parseImports(
+                    `import "lib";`,
+                    `import { x } from "lib";`);
+                const actualCoalescedImports = OrganizeImports.coalesceImports(sortedImports);
+                const expectedCoalescedImports = sortedImports;
+                assertListEqual(expectedCoalescedImports, actualCoalescedImports);
+            });
+
+            it("Combine namespace import with default import", () => {
+                const sortedImports = parseImports(
```
[the remark this answers is not in the document]
Comment (amcasey) [reply src/harness/unittests/organizeImports.ts]: I left it expanded because I thought it made the test more readable.  A list of unnamed strings can be hard to follow.

File: src/harness/unittests/organizeImports.ts [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+
+            function runBaseline(baselinePath: string, testFile: TestFSWithWatch.FileOrFolder, ...otherFiles: TestFSWithWatch.FileOrFolder[]) {
+                const { path: testPath, content: testContent } = testFile;
+                const languageService = makeLanguageService(testFile, ...otherFiles);
+                const changes = languageService.organizeImports({ type: "file", fileName: testPath }, testFormatOptions);
+                assert.equal(1, changes.length);
+                assert.equal(testPath, changes[0].fileName);
+
+                Harness.Baseline.runBaseline(baselinePath, () => {
+                    const data: string[] = [];
```
[the remark this answers is not in the document]
Comment (amcasey) [reply src/harness/unittests/organizeImports.ts]: Frankly, I didn't understand why we didn't just build a string directly.  Is the array literal preferable?

File: src/services/organizeImports.ts [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+        return isStringLiteral(specifier) || isNoSubstitutionTemplateLiteral(specifier)
+            ? specifier.text
+            : undefined;
+    }
+
+    /**
+     * @param sortedImports a non-empty list of ImportDeclarations, sorted by module name.
+     */
+    function groupSortedImports(sortedImports: ReadonlyArray<ImportDeclaration>): ReadonlyArray<ReadonlyArray<ImportDeclaration>> {
+        Debug.assert(length(sortedImports) > 0);
```
[the remark this answers is not in the document]
Comment (andy-ms) [reply src/services/organizeImports.ts]: Maybe better:
```ts
function groupSortedImports(sortedImports: ReadonlyArray<ImportDeclaration>): ReadonlyArray<ReadonlyArray<ImportDeclaration>> {
    return groupBy(sortedImports, (a, b) => getExternalModuleName(a.moduleSpecifier) === getExternalModuleName(b.moduleSpecifier));
}

function groupBy<T>(values: ReadonlyArray<T>, areSameGroup: (a: T, b: T) => boolean): ReadonlyArray<ReadonlyArray<T>> {
    Debug.assert(values.length !== 0);
    const groups: T[][] = [];
    let group: T[] = [];
    for (const value of values) {
        if (group.length && !areSameGroup(value, group[0])) {
            groups.push(group);
            group = [];
        }
        group.push(value);
    }
    if (group.length) {
        groups.push(group);
    }
    return groups;
}
```

File: src/services/organizeImports.ts [hunk 15]
```diff
@@ -0,0 +1,291 @@
+/* @internal */
+namespace ts.OrganizeImports {
+    export function organizeImports(
+        sourceFile: SourceFile,
+        formatContext: formatting.FormatContext,
+        host: LanguageServiceHost,
+        cancellationToken: CancellationToken) {
+
+        // All of the (old) ImportDeclarations in the file, in syntactic order.
+        const oldImportDecls: ImportDeclaration[] = [];
```
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/organizeImports.ts]: Fair.

File: src/harness/unittests/organizeImports.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (andy-ms) [reply src/harness/unittests/organizeImports.ts]: Well, it lets you avoid writing `${newLineCharacter}` a lot... though that's hardcoded to `\n` anyway, so you could just write `\n`.

File: src/services/organizeImports.ts [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+
+        const usedImportDecls = removeUnusedImports(oldImportDecls);
+        cancellationToken.throwIfCancellationRequested();
+        const sortedImportDecls = sortImports(usedImportDecls);
+        cancellationToken.throwIfCancellationRequested();
+        const coalescedImportDecls = coalesceImports(sortedImportDecls);
+        cancellationToken.throwIfCancellationRequested();
+
+        // All of the (new) ImportDeclarations in the file, in sorted order.
+        const newImportDecls = coalescedImportDecls;
```
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/organizeImports.ts]: Are locals expensive in TS?  I like naming things - it makes code easier to read.

File: src/services/organizeImports.ts [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+            // Delete the surrounding trivia because it will have been retained in newImportDecls.
+            const replaceOptions = {
+                useNonAdjustedStartPosition: false,
+                useNonAdjustedEndPosition: false,
+                suffix: getNewLineOrDefaultFromHost(host, formatContext.options),
+             };
+            changeTracker.replaceNodeWithNodes(sourceFile, oldImportDecls[0], newImportDecls, replaceOptions);
+        }
+
+        const changes = changeTracker.getChanges();
```
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/organizeImports.ts]: Sure

File: src/services/organizeImports.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (andy-ms) [reply src/services/organizeImports.ts]: I'm sure modern JS engines know that these are the same thing. I normally agree that giving a descriptive name for a variable is useful, but `const coalescedImportDecls = coalesceImports(sortedImportDecls);` just repeats the same information in two different names, since the function name already tells you what it's doing -- same for the above lines.

File: src/services/organizeImports.ts (hunk 5 above)
Comment (amcasey) [reply src/services/organizeImports.ts]: I'm not sure I understand your concerns about the current implementation.

File: src/services/organizeImports.ts [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+
+        return groups;
+    }
+
+    /* @internal */ // Internal for testing
+    /**
+     * @param sortedImports a list of ImportDeclarations, sorted by module name.
+     */
+    export function coalesceImports(sortedImports: ReadonlyArray<ImportDeclaration>) {
+        if (sortedImports.length === 0) {
```
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/organizeImports.ts]: I believe this also runs after unused import removal.

File: src/services/organizeImports.ts (hunk 5 above)
Comment (andy-ms) [reply src/services/organizeImports.ts]: It's not that I spotted any bugs, just noticed that it could be done in 4 lines.

File: src/services/organizeImports.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/organizeImports.ts]: In the debugger, it's easy to inspect each stage individually.  If you'd prefer not to have any redundant text, arguably they should be combined into a single expression with only final value stored in a named variable.

File: src/services/organizeImports.ts [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+        for (let i = oldImportDecls.length - 1; i > 0; i--) {
+            changeTracker.deleteNode(sourceFile, oldImportDecls[i]);
+        }
+
+        if (newImportDecls.length === 0) {
+            changeTracker.deleteNode(sourceFile, oldImportDecls[0]);
+        }
+        else {
+            // Delete the surrounding trivia because it will have been retained in newImportDecls.
+            const replaceOptions = {
```
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/organizeImports.ts]: Sure.

File: src/services/organizeImports.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (andy-ms) [reply src/services/organizeImports.ts]: For pure functions, you can enter e.g. `coalesceImports(sortedImportDecls)` into the watch list for inspection without needing a named local variable.

File: src/services/organizeImports.ts (hunk 5 above)
Comment (amcasey) [reply src/services/organizeImports.ts]: It hadn't occurred to me to combine `Comparison`s with `||` - applying that should clean things up a bit.

One of the reasons I pulled out the struct was so that the same work wouldn't be repeated for every comparison in which a given import participated.  If that's not interesting, it can obviously be simplified.

File: src/services/organizeImports.ts [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+        const coalescedImportDecls = coalesceImports(sortedImportDecls);
+        cancellationToken.throwIfCancellationRequested();
+
+        // All of the (new) ImportDeclarations in the file, in sorted order.
+        const newImportDecls = coalescedImportDecls;
+
+        const changeTracker = textChanges.ChangeTracker.fromContext({ host, formatContext });
+
+        // NB: Stopping before i === 0
+        for (let i = oldImportDecls.length - 1; i > 0; i--) {
```
Comment (andy-ms) [inline src/services/organizeImports.ts]: Could you add a comment explaining why this is in reverse order? I assume this has something to do with trivia?

File: src/services/organizeImports.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/organizeImports.ts]: Just to confirm, your improvement was pulling out a generic helper?  The actual behavior is the same?

File: src/services/organizeImports.ts (hunk 8 above)
Comment (amcasey) [reply src/services/organizeImports.ts]: Sure.

File: src/services/organizeImports.ts (hunk 7 above)
Comment (amcasey) [reply src/services/organizeImports.ts]: Thanks, I figured there had to be a function that did that.  I was looking for "selectMany".

File: src/services/organizeImports.ts (hunk 5 above)
Comment (andy-ms) [reply src/services/organizeImports.ts]: If you filter out invalid imports at the top, `getExternalModuleName` is just a property lookup. `isExternalModuleNameRelative` isn't very complicated either -- some simple regex and `indexOf` tests.

File: src/services/organizeImports.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (andy-ms) [reply src/services/organizeImports.ts]: Yes, assuming invalid imports are filtered out (so `getExternalModuleName`) returns a defined result).

[the remark this answers is not in the document]
Comment (andy-ms) [reply src/services/organizeImports.ts]: Just noticed we don't need `Debug.assert(values.length !== 0);` in the "maybe better" version.

File: src/services/organizeImports.ts [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+                    : updateNamedImports(namedImports[0], sortedImportSpecifiers);
+
+            coalescedImports.push(
+                updateImportDeclarationAndClause(importClause, newDefaultImport, newNamedImports));
+        }
+
+        return coalescedImports;
+
+        // `undefined` is the min value.
+        function compareIdentifiers(s1: Identifier | undefined, s2: Identifier | undefined) {
```
[the remark this answers is not in the document]
Comment (andy-ms) [reply src/services/organizeImports.ts]: Opinions on that vary around the team, so whichever is fine.

File: src/harness/unittests/organizeImports.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (RyanCavanaugh) [reply src/harness/unittests/organizeImports.ts]: In older JS runtimes before everyone switched to ropes, repeated string concat was expensive. `string[]` -> one `string` via `join` is basically the equivalent of using a `StringBuilder` in C#

File: src/services/organizeImports.ts (hunk 20 above)
Comment (amcasey) [reply src/services/organizeImports.ts]: It's not - I was thinking of something else.

File: src/services/organizeImports.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/organizeImports.ts]: Note that this isn't true group-by - it requires that the input be sorted.  Is it still worth pulling out a generic helper for specialized functionality?

[the remark this answers is not in the document]
Comment (andy-ms) [reply src/services/organizeImports.ts]: Actually, this doesn't *just* rely on it being sorted, but on it being sorted by the same mechanism as `areSameGroup`. Maybe it would be better to group first (not relying on sorting) and then sort the groups?

File: src/services/organizeImports.ts [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+
+        // All of the old ImportDeclarations in the file, in syntactic order.
+        const oldImportDecls = sourceFile.statements.filter(isImportDeclaration);
+
+        if (oldImportDecls.length === 0) {
+            return [];
+        }
+
+        const oldValidImportDecls = oldImportDecls.filter(importDecl => getExternalModuleName(importDecl.moduleSpecifier));
+        const oldInvalidImportDecls = oldImportDecls.filter(importDecl => !getExternalModuleName(importDecl.moduleSpecifier));
```
Comment (andy-ms) [inline src/services/organizeImports.ts]: Not sure this is needed -- we're basically just deleting these and adding them back, could just ignore them entirely and make `oldImportDecls` only be the valid ones.

File: src/services/organizeImports.ts [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+            const name2 = getExternalModuleName(import2.moduleSpecifier);
+            Debug.assert(name1 !== undefined);
+            Debug.assert(name2 !== undefined);
+            return compareBooleans(isExternalModuleNameRelative(name1), isExternalModuleNameRelative(name2)) ||
+                compareStringsCaseSensitive(name1, name2);
+
+        });
+    }
+
+    function getExternalModuleName(specifier: Expression) {
```
Comment (andy-ms) [inline src/services/organizeImports.ts]: Could have a `ValidImportDeclaration` type that always has a valid specifier, and then not need to use this function or worry about `undefined` results.
```ts
interface ValidImportDeclaration extends ImportDeclaration { moduleSpecifier: StringLiteral; }
function isValidImportDeclaration(node: Node): node is ValidImportDeclaration {
    return isImportDeclaration(node) && isStringLiteral(node.moduleSpecifier);
}
```
then:
```ts
const oldImportDecls = sourceFile.statements.filter(isValidImportDeclaration);
```
(We could technically allow `NoSubstitutionTemplateLiteral`s here, but they're actually compile errors so not really valid.)

File: src/services/organizeImports.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/organizeImports.ts]: Yes, that's why it's a private helper function.  At the moment, there doesn't seem to be a need for something more general, does there?

File: src/services/organizeImports.ts (hunk 22 above)
Comment (amcasey) [reply src/services/organizeImports.ts]: If they are scattered throughout the file, this will group them together at the top.

Comment (andy-ms) [reply src/services/organizeImports.ts]: OK. For the case where you do keep the invalid ones, could use a helper function and write `const [oldValidImportDecls, oldInvalidImportDecls] = partition(oldImportDecls, isValidImportDeclaration);`

File: src/services/organizeImports.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/organizeImports.ts]: As you requested, we are now grouping before performing the other operations.

**Comment by amcasey** [comment]
Restarting tests since lint error line numbers didn't match current source.

**Comment by amcasey** [comment]
I am unable to repro the failure `The baseline file organizeImports/CoalesceMultipleModules.ts has changed`.

**Comment by andy-ms** [comment]
@amcasey Might have to do with newlines?

**Comment by amcasey** [comment]
@andy-ms Conceivably, but it would be strange to have only this one failure.

**Comment by amcasey** [comment]
We have many files with mixed line endings, but my current hypothesis is that the problem was introduced when I renamed the baseline output directory from "OrganizeImports" to "organizeImports" for consistency with the others.  I'll try to purge the old name locally and then force push.

**Comment by amcasey** [comment]
Other than this irritating issue, is this more or less ready to merge?

File: src/server/protocol.ts [hunk 24]
```diff
@@ -547,6 +551,27 @@ namespace ts.server.protocol {
         renameFilename?: string;
     }
 
+    /**
```
Comment (andy-ms) [inline src/server/protocol.ts]: This comment might be better in `organizeImports.ts` (the non-test one).

**Comment by amcasey** [comment]
I've revised history so that the baseline folder has always been lowercase.  Hopefully, that will fix the tests.

File: src/harness/unittests/organizeImports.ts (hunk 3 above)
Comment (andy-ms) [reply src/harness/unittests/organizeImports.ts]: No, baseline tests will do.

File: src/server/protocol.ts (hunk 24 above)
Comment (amcasey) [reply src/server/protocol.ts]: I can move it when I implement remove-unused.

**Comment by MgSam** [comment]
How do you invoke this command in Visual Studio? I've been completely unable to get it to appear anywhere since upgrading to 2.8.

**Comment by mhegazy** [comment]
This is a new command, so requires update to VS as well. Should light up in your next VS update.
